File: src/content/3.4/code/scala/snippet02.scala
case class Writer[W, A](run: (A, W))

implicit def writerFunctor[W] = new Functor[Writer[W, ?]] {
  def fmap[A, B](f: A => B)(fa: Writer[W, A]): Writer[W, B] =
    fa match {
      case Writer((a, w)) => Writer(f(a), w)
  }
}

File: src/content/3.4/code/ocaml/snippet02.ml
module WriterInstance (W : sig
  type w
end) : Functor with type 'a t = (W.w, 'a) writer = struct
  type 'a t = (W.w, 'a) writer

  let fmap f (Writer (a, w)) = Writer (f a, w)
end
